### DISCUSSION TITLE
请教:通过StreamWriter将字符串写入ini文档,但编码格式不对

### DISCUSSION THREAD

**Discussion by RanZhao233** [BYJRK/DotNet-Discussions#251] [C# 技术交流]
框架是 : Framework4.7.2
代码如下:
using (StreamWriter writer = new StreamWriter(strFileName, false, Encoding.GetEncoding("GB2312")))
{
    StringBuilder sb = new StringBuilder(); 
    string a = "[Global]";
    string b = "1=" + txt.Text;
    string c = "[1]";
    string d = "ChangeData=";
    string e = "Cnt=0";
    string f = "IsShow=0";
    string g = "IsTemplate=1";
    sb.AppendLine(a);
    sb.AppendLine(b);
    sb.AppendLine(c);
    sb.AppendLine(d);
    sb.AppendLine(e);
    sb.AppendLine(f);
    sb.AppendLine(g);

    // 将内容写入文件
    writer.Write(sb.ToString());
}

我指定了GB2312,但写入ini文档以后,用notepad++打开,确实UTF8编码.
更诡异的是,如果我的txt.Text是"水",那么百分百是UTF8编码,如果是其他的中文,有一定概率是ANSI编码.
我的需求是写入文档的内容能正常显示,并且文档编码格式就是ANSI的,同时想明白为什么会出现这么情况.
恳请各位大佬不吝赐教!万分感谢!
![image](https://github.com/BYJRK/DotNet-Discussions/assets/155596269/27f353d3-eb90-4767-96ed-7c09ae4ebd1b)
![image](https://github.com/BYJRK/DotNet-Discussions/assets/155596269/9de56dbe-1dff-4cf0-a9be-ea74fe157782)

**Comment by BYJRK** [comment]
你是不是没有配置 CodePages？新版本的 .NET 默认已经不再包含 Unicode 以外的编码了。需要安装 `System.Text.Encoding.CodePages`，然后 `Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);`

**Reply by RanZhao233** [reply]
没想到得到本尊回复了!我使用的是Framework4.7.2,然后刚才也尝试安装System.Text.Encoding.CodePages,并且注册了.还是不行.
并且在我 new StreamWriter的时候,目标文件已经创建,用NPP打开是ANSI编码的,然后writer.Write(sb.ToString());这句执行完毕以后,再打开,就变成UTF-8编码格式了.

**Reply by BYJRK** [reply]
如果你输出的是中文，然后被notepad++识别成了UTF8，那么显示的是不是乱码？

**Reply by RanZhao233** [reply]
没有呀,文本文档一开始都是不存在的.是新建的. 我在StringBuilder sb = new StringBuilder();打了断点,也就是using (StreamWriter writer = new StreamWriter(strFileName, false, Encoding.GetEncoding("GB2312")))的时候新建了ini文件,我用npp启动,可以看到是ANSI编码的. 然后等最后writer.Write(sb.ToString());执行完以后,我再用npp打开这个文档,编码就变成utf-8了. "水"字就会变成问号,如果我手动切换为ANSI是可以看到水字的.
并且很诡异的地方是,并不是所有的中文都会导致这种现象.当我在xmal里的 txt输入其他中文时,最后导出的文本有概率是正常的.我有测试过还有"血""全""水"这些字,百分百会出问题.
主要还是对这种现象感到很困惑.

**Reply by RanZhao233** [reply]
我手动创建这样的文档,以ANSI编码保存![image](https://github.com/BYJRK/DotNet-Discussions/assets/155596269/0c87ce70-d5d3-40a2-b34c-bf825729bfd4). 重新打开以后npp也是把它识别的utf8编码~感觉应该不是代码的原因,是文本格式的原因

**Comment by SlimeNull** [comment]
编码猜测问题吧。我觉得是Notepad++没能正确猜出来文件编码，你可以用别的编辑器打开试试，看它们能否猜出正确的编码。
